### PR TITLE
ExactSizeIterator

### LINKED ISSUE #46: Return `ExactSizeIterator` from shapes where possible
https://github.com/ManevilleF/hexx/blob/9e7dc161e219e8d3147b4b1d3fc461504c14fe62/src/lib.rs#L156

This avoids needing to slowly count the iterator. This was particularly silly because iterators are consumed by doing so.

See this code that I wrote:

```rust
    /// Returns the average height of tiles around `tile_pos` within `radius`
    pub(crate) fn average_height(&self, tile_pos: TilePos, radius: u32) -> f32 {
        let hex_iter = hexagon(tile_pos.hex, radius);
        let n = hex_iter.count();
        let hex_iter = hexagon(tile_pos.hex, radius);
        let heights = hex_iter.map(|hex| *self.height_index.get(&TilePos { hex }).unwrap_or(&0.));
        heights.sum::<f32>() / n as f32
    }
```

### PR DESCRIPTION
**Pull request by ManevilleF**
> Closes #46 

# Work Done

- [x] Added a private iterator wrapper `ExactSizeHexIterator` 
- [x] Wrapped known size iterator methods in the wrapper:
  - [x] `Hex::range`
  - [x] `Hex::line_to`
  - [x] `shapes::hexagon`  
  - [x] `HexBounds::all_coords` 
 - [x] Extend the same for rings
   - [x] ring edges 
   - [x] Added *full wedges* with exact size 
 - [x] Extend the same for other shapes where applicable

### CODE REVIEW

File: src/hex/iter.rs [hunk 1]
```diff
@@ -66,3 +66,30 @@ impl<I: Iterator<Item = Hex>> HexIterExt for I {
         self.collect()
     }
 }
+
+/// Private container for a [`Hex`] [`Iterator`] of known size
+#[derive(Debug, Clone)]
+pub struct ExactSizeHexIterator<I> {
```
Comment (alice-i-cecile) [inline src/hex/iter.rs:72]: The fact that this is `pub` is rather confusing to me, given that this is described as being private.

File: src/hex/rings.rs [hunk 2]
```diff
@@ -175,11 +179,11 @@ impl Hex {
     /// If you only need the coordinates see [`Self::wedge`]
     pub fn custom_ring_edges(
         self,
-        range: impl Iterator<Item = u32>,
+        ranges: impl Iterator<Item = u32>,
         direction: DiagonalDirection,
         clockwise: bool,
     ) -> impl Iterator<Item = Vec<Self>> {
-        range.map(move |r| self.custom_ring_edge(r, direction, clockwise))
+        ranges.map(move |r| self.custom_ring_edge(r, direction, clockwise).collect())
```
Comment (alice-i-cecile) [inline src/hex/rings.rs:186]: There's no need to collect with the current set of changes.

File: src/hex/iter.rs (hunk 1 above)
Comment (ManevilleF) [reply src/hex/iter.rs:72]: It is in a private module, innaccessible from library users

File: src/hex/rings.rs (hunk 2 above)
Comment (ManevilleF) [reply src/hex/rings.rs:186]: Then the return value would need to change to `impl Iterator<Item = impl ExactSizeIterator<Item = Self>>`

Comment (ManevilleF) [reply src/hex/rings.rs]: Seems to be working fine

**Pull request merged**